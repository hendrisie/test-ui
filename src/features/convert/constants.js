export const NAME = 'convert';

export const ACTION_SUBMIT = 'features/convert/ACTION_SUBMIT';
