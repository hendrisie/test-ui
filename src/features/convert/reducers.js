/*
@startuml

class convert.reducers << (R,green) Reducer>>{
    +denomination: Array<number>
    +result: Array<denomination>,
    +leftover: number,
}

@enduml

 */
import { ACTION_SUBMIT } from './constants';
import { parser } from './formhelper';

export const initialState = {
  denomination: [100000, 50000, 20000, 10000, 5000, 1000, 500, 100, 50],
  result: null,
  leftover: null
};

export const calculateChange = (totalVal, denomination) => {
  // const denomination = [100000,50000,20000,10000,5000,1000,500,100,50];
  // sanitize input from action.payload
  let total = parser.currencyParser(totalVal);
  // sort through denomination descendingly
  let denominators = denomination ? denomination : [];
  let leftover = 0;
  const result = [];
  // iterate through all possible denominators from the highest to lowest
  for (const denom of denominators) {
    if (total >= denom) {
      // find how many qty needed for current denominator
      const qty = Math.floor(total / denom);
      result.push([qty, denom]);
      // update total with the remainder value
      total = total % denom;
      // if no more remainder, exit loop
      if (total === 0) break;
    }
  }
  // whatever left from total is updated to leftover variable
  if (total > 0) leftover = total;
  return { result, leftover };
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ACTION_SUBMIT:
      if (action.payload) {
        const { result, leftover } = calculateChange(action.payload, state.denomination);
        return {
          ...state,
          result,
          leftover
        };
      }
      return state;
    default:
      return state;
  }
};

export default reducer;
