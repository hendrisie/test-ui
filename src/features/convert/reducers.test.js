import reducer from './reducers';
import { initialState } from './reducers';
import { ACTION_SUBMIT, NAME } from './constants';

describe('convert reducer', () => {
  let state;
  beforeEach(() => {
    state = initialState;
  });
  it('should return the initial state', () => {
    expect(reducer(undefined, {})).toMatchSnapshot();
  });

  it('should handle ACTION_SUBMIT with proper payload', () => {
    expect(
      reducer(state, {
        type: ACTION_SUBMIT,
        payload: 'Rp 100500'
      })
    ).toMatchSnapshot();
  });

  it('should not error if empty payload', () => {
    expect(
      reducer(state, {
        type: ACTION_SUBMIT,
        payload: ''
      })
    ).toMatchSnapshot();
  });

  it('should not error when null payload and empty denomination', () => {
    state = {
      denomination: null,
      ...initialState
    };
    expect(
      reducer(state, {
        type: ACTION_SUBMIT,
        payload: null
      })
    ).toMatchSnapshot();
  });
});
