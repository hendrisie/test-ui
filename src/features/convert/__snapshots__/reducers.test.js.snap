// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convert reducer should handle ACTION_SUBMIT with proper payload 1`] = `
Object {
  "denomination": Array [
    100000,
    50000,
    20000,
    10000,
    5000,
    1000,
    500,
    100,
    50,
  ],
  "leftover": 0,
  "result": Array [
    Array [
      1,
      100000,
    ],
    Array [
      1,
      500,
    ],
  ],
}
`;

exports[`convert reducer should not error if empty payload 1`] = `
Object {
  "denomination": Array [
    100000,
    50000,
    20000,
    10000,
    5000,
    1000,
    500,
    100,
    50,
  ],
  "leftover": null,
  "result": null,
}
`;

exports[`convert reducer should not error when null payload and empty denomination 1`] = `
Object {
  "denomination": Array [
    100000,
    50000,
    20000,
    10000,
    5000,
    1000,
    500,
    100,
    50,
  ],
  "leftover": null,
  "result": null,
}
`;

exports[`convert reducer should return the initial state 1`] = `
Object {
  "denomination": Array [
    100000,
    50000,
    20000,
    10000,
    5000,
    1000,
    500,
    100,
    50,
  ],
  "leftover": null,
  "result": null,
}
`;
